KeyVault.Acmebot/Providers/TransIpProvider.cs

```diff
@@ -0,0 +1,333 @@
+        }
+
+        private readonly TransIpClient _transIpClient;
+
+        public int PropagationSeconds => 360;
+
+        public async Task CreateTxtRecordAsync(DnsZone zone, string relativeRecordName, IEnumerable<string> values)
+        {
+            var records = values.Select(value => new DnsEntry()
+            {
+                Name = relativeRecordName,
+                Type = "TXT",
+                Expire = 60,
+                Content = value
+            });
+
+            foreach (var record in records)
+                await _transIpClient.AddRecord(zone.Name, record);
+        }
+
+        public async Task DeleteTxtRecordAsync(DnsZone zone, string relativeRecordName)
+        {
+            var records = await _transIpClient.ListRecords(zone.Name);
+
+            var recordsToDelete = records.Where(r => string.Equals(relativeRecordName, r.Name) && r.Type == "TXT");
+
+            foreach (var record in recordsToDelete)
+                await _transIpClient.DeleteRecord(zone.Name, record);
+        }
+
+        public async Task<IReadOnlyList<DnsZone>> ListZonesAsync()
+        {
+            var zones = await _transIpClient.ListZonesAsync();
+
+            return zones.Select(d => new DnsZone() { Id = d.Name, Name = d.Name }).ToArray();
+        }
+
+        private class TransIpClient
+        {
+            public TransIpClient(string customerName, CryptographyClient cryptoClient)
+            {
+                _customerName = customerName;
+                _cryptoClient = cryptoClient;
+
+                _httpClient = new HttpClient()
+                {
+                    BaseAddress = new Uri("https://api.transip.nl/v6/")
+                };
+
+                _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            }
+
+            private readonly HttpClient _httpClient;
+            private readonly string _customerName;
+            private readonly CryptographyClient _cryptoClient;
+            private TransIPToken _token = null;
+
+            public async Task<IEnumerable<Domain>> ListZonesAsync()
+            {
+                await EnsureLoggedInAsync();
+
+                var response = await _httpClient.GetAsync("domains");
+
+                response.EnsureSuccessStatusCode();
+
+                var domains = await response.Content.ReadAsAsync<ListDomainsResult>();
+
+                return domains.Domains;
+            }
+
+            public async Task<IReadOnlyList<DnsEntry>> ListRecords(string zoneName)
+            {
+                await EnsureLoggedInAsync();
+
+                HttpResponseMessage response = await _httpClient.GetAsync($"domains/{zoneName}/dns");
+
+                response.EnsureSuccessStatusCode();
+
+                ListDnsEntriesResponse entries = await response.Content.ReadAsAsync<ListDnsEntriesResponse>();
+
+                return entries.DnsEntries.ToList();
```

Comment:
```suggestion
                return entries.DnsEntries;
```

Change the model class to `IReadOnlyList<T>` so that `ToArray` is no longer needed.